src/supernova/mod.rs

```diff
@@ -0,0 +1,1202 @@
+
+    self.r_U_primary[claim.get_augmented_circuit_index()]
+      .as_ref()
+      .map_or(Ok(()), |U| {
+        if U.X.len() != 2 {
+          println!("r_U_primary got instance length {:?} != {:?}", U.X.len(), 2);
+          Err(NovaError::ProofVerifyError)
+        } else {
+          Ok(())
+        }
+      })?;
+
+    self.r_U_secondary[0].as_ref().map_or(Ok(()), |U| {
+      if U.X.len() != 2 {
+        println!(
+          "r_U_secondary got instance length {:?} != {:?}",
+          U.X.len(),
+          2
+        );
+        Err(NovaError::ProofVerifyError)
+      } else {
+        Ok(())
+      }
+    })?;
+
+    // check if the output hashes in R1CS instances point to the right running instances
+    let num_field_primary_ro = 3 // params_next, i_new, program_counter_new
+    + 2 * pp.F_arity_primary // zo, z1
+    + (7 + 2 * pp.augmented_circuit_params_primary.get_n_limbs()); // # 1 * (7 + [X0, X1]*#num_limb)
+    let num_field_secondary_ro = 3 // params_next, i_new, program_counter_new
+    + 2 * pp.F_arity_primary // zo, z1
+    + self.num_augmented_circuits * (7 + 2 * pp.augmented_circuit_params_primary.get_n_limbs()); // #num_augmented_circuits * (7 + [X0, X1]*#num_limb)
+
+    let (hash_primary, hash_secondary) = {
+      let mut hasher = <G2 as Group>::RO::new(pp.ro_consts_secondary.clone(), num_field_primary_ro);
+      hasher.absorb(self.pp_digest);
+      hasher.absorb(G1::Scalar::from(self.i as u64));
+      hasher.absorb(self.program_counter);
+
+      for e in z0_primary {
+        hasher.absorb(*e);
+      }
+      for e in &self.zi_primary {
+        hasher.absorb(*e);
+      }
+      if let Some(U) = self.r_U_secondary[0].as_ref() {
+        U.absorb_in_ro(&mut hasher)
+      }
+
+      let mut hasher2 =
+        <G1 as Group>::RO::new(pp.ro_consts_primary.clone(), num_field_secondary_ro);
+      hasher2.absorb(scalar_as_base::<G1>(self.pp_digest));
+      hasher2.absorb(G2::Scalar::from(self.i as u64));
+      hasher2.absorb(G2::Scalar::ZERO);
+      for e in z0_secondary {
+        hasher2.absorb(*e);
+      }
+      for e in &self.zi_secondary {
+        hasher2.absorb(*e);
+      }
+      let default_value = RelaxedR1CSInstance::default(ck_primary, &pp.r1cs_shape_primary);
+      self.r_U_primary.iter().for_each(|U| {
+        U.as_ref()
+          .unwrap_or(&default_value)
+          .absorb_in_ro(&mut hasher2);
+      });
+
+      (
+        hasher.squeeze(NUM_HASH_BITS),
+        hasher2.squeeze(NUM_HASH_BITS),
+      )
+    };
+
+    if hash_primary != self.l_u_secondary.X[0] {
+      println!(
+        "hash_primary {:?} not equal l_u_secondary.X[0] {:?}",
+        hash_primary, self.l_u_secondary.X[0]
+      );
+      return Err(NovaError::ProofVerifyError);
+    }
+    if hash_secondary != scalar_as_base::<G2>(self.l_u_secondary.X[1]) {
+      println!(
+        "hash_secondary {:?} not equal l_u_secondary.X[1] {:?}",
+        hash_secondary, self.l_u_secondary.X[1]
+      );
```

Comment:
The `println!` statements are very understandable (I've inserted similar ones in that position in my local branches myself), but we probably don't want them here. Consider using the `log` crate, with a `debug` log-level instead.

Reply:
Tried to switch to log crate in commit https://github.com/microsoft/Nova/pull/204/commits/61dad5b531b7316313b991d9651a32584286a6c6